### PR TITLE
[gql][consistent-reads][7/n] Extend Balance cursor to include checkpoint_viewed_at

### PR DESCRIPTION
**Pull request by wlmyng**
## Description 

Even though `Balance` does not resolve to other non-Scalar nodes, because there is a `Balance` connection, we will still need to track `checkpoint_viewed_at`. This is so that the caller stays on the same checkpoint while paginating through a balance connection.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 1]
```diff
@@ -40,7 +41,15 @@ pub struct StoredBalance {
     pub coin_type: String,
 }
 
-pub(crate) type Cursor = cursor::JsonCursor<String>;
+pub(crate) type Cursor = cursor::JsonCursor<BalanceCursor>;
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/balance.rs:44]: nit: blank line after type decl.

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 2]
```diff
@@ -142,7 +157,17 @@ impl RawPaginated<Cursor> for StoredBalance {
 
 impl Target<Cursor> for StoredBalance {
     fn cursor(&self) -> Cursor {
-        Cursor::new(self.coin_type.clone())
+        Cursor::new(BalanceCursor {
+            coin_type: self.coin_type.clone(),
+            checkpoint_viewed_at: None,
+        })
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/balance.rs:163]: Do we expect this path to get used anymore? If not, should we `panic!` in it to make sure it doesn't get used?

File: crates/sui-graphql-rpc/src/types/balance.rs [hunk 3]
```diff
@@ -249,3 +274,24 @@ fn filter(mut query: RawQuery, owner: SuiAddress, coin_type: Option<TypeTag>) ->
 
     query
 }
+
+/// Check that the cursors, if provided, have the same checkpoint_viewed_at.
+pub(crate) fn validate_cursor_consistency(
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/balance.rs:279]: Worth thinking about a neat way to abstract this, seeing as it's quite a common pattern. Sketch for that:

- Add a trait: `ConsistentCursor: CursorType` that exposes the `checkpoint_viewed_at`.
- Split `Target<C>` into `Target<C>` (as it was before) and `ConsistentTarget<C: ConsistentCursor>: Target<C>` that exposes the `consistent_cursor` builder.
- Add an impl block: `impl<C: ConsistentCursor> Page<C> { fn validate_consistency(&self) -> Option<u64> }`.

**Review by amnn — APPROVED** [review]
LGTM!

File: crates/sui-graphql-rpc/src/types/balance.rs (hunk 3 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/balance.rs:279]: Applied these concepts later in the stack at https://github.com/MystenLabs/sui/pull/15996

File: crates/sui-graphql-rpc/src/types/balance.rs (hunk 2 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/balance.rs:163]: Handled at https://github.com/MystenLabs/sui/pull/15996

**Pull request merged**